### PR TITLE
Add types to public and protected properties

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | 7.0
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | -
| License       | MIT
| Doc PR        | -

:hot_face:  

Allowed by https://github.com/symfony/symfony/pull/45360

Follows #51068 and #51067

### CODE REVIEW

File: src/Symfony/Bundle/FrameworkBundle/Translation/Translator.php
Comment (94noni) [inline src/Symfony/Bundle/FrameworkBundle/Translation/Translator.php:31]: passing by, are those removal desired or code swap typo?

File: src/Symfony/Component/Config/Definition/BaseNode.php
Comment (derrabus) [inline src/Symfony/Component/Config/Definition/BaseNode.php]:
```suggestion
    protected string $name;
```

The constructor makes sure that `$name` is not `null`.

https://github.com/symfony/symfony/blob/f10050724a1d2b5a9d57a30aa730be487d620283/src/Symfony/Component/Config/Definition/BaseNode.php#L51

File: src/Symfony/Bundle/FrameworkBundle/Controller/AbstractController.php
Comment (derrabus) [inline src/Symfony/Bundle/FrameworkBundle/Controller/AbstractController.php:58]: `setContainer()` will read this property _before_ writing it. Thus, we have to declare it nullable and initialize it with `null`.

Alternatively, we could adjust the access in `setContainer()`:

```php
$previous = $this->container ?? null;
```

File: src/Symfony/Bundle/FrameworkBundle/Test/KernelTestCase.php
Comment (derrabus) [inline src/Symfony/Bundle/FrameworkBundle/Test/KernelTestCase.php]:
```suggestion
    protected static ?KernelInterface $kernel = null;
```

Otherwise `ensureKernelShutdown()` will trip an uninitialized property error if the kernel hasn't been created.

File: src/Symfony/Component/BrowserKit/Cookie.php
Comment (derrabus) [inline src/Symfony/Component/BrowserKit/Cookie.php]: I don't think this property should be nullable.

File: src/Symfony/Component/Cache/Adapter/DoctrineDbalAdapter.php
Comment (derrabus) [inline src/Symfony/Component/Cache/Adapter/DoctrineDbalAdapter.php]: Psalm has a point here: In `AbstractAdapterTrait` this property is declared nullable. Maybe we should stop overriding this property and move the assignment to the constructor?

File: src/Symfony/Component/Cache/Adapter/MemcachedAdapter.php
Comment (derrabus) [inline src/Symfony/Component/Cache/Adapter/MemcachedAdapter.php]: same

File: src/Symfony/Component/Cache/Adapter/PdoAdapter.php
Comment (derrabus) [inline src/Symfony/Component/Cache/Adapter/PdoAdapter.php]: same

File: src/Symfony/Component/HttpFoundation/StreamedResponse.php
Comment (derrabus) [inline src/Symfony/Component/HttpFoundation/StreamedResponse.php:29]: `setCallback()` accepts any callable. You need to make sure we always assign a closure:

```php
public function setCallback(callable $callback): static
{
    $this->callback = $callback(...);
    return $this;
}
```

In addition to this, in `sendContent()` we check if `$this->callback` is null.

https://github.com/symfony/symfony/blob/f10050724a1d2b5a9d57a30aa730be487d620283/src/Symfony/Component/HttpFoundation/StreamedResponse.php#L90-L92

We might want to change this to an `isset()` check, I suppose.

**Review by derrabus — CHANGES_REQUESTED** [review]
Reviewing this PR is exhausting. I've reviewed up to the HttpFoundation component and will take a break now. 😓

File: src/Symfony/Bundle/FrameworkBundle/Translation/Translator.php
Comment (nicolas-grekas) [reply src/Symfony/Bundle/FrameworkBundle/Translation/Translator.php:31]: fixed

File: src/Symfony/Bundle/FrameworkBundle/Controller/AbstractController.php
Comment (nicolas-grekas) [reply src/Symfony/Bundle/FrameworkBundle/Controller/AbstractController.php:58]: alternative applied

File: src/Symfony/Bundle/FrameworkBundle/Routing/Router.php
Comment (derrabus) [inline src/Symfony/Bundle/FrameworkBundle/Routing/Router.php]: For 6.4? Otherwise FrameworkBundle 6.4 won't be compatible with Routing 7 once this PR is merged.

File: src/Symfony/Component/Config/Definition/Builder/NodeBuilder.php
Comment (derrabus) [inline src/Symfony/Component/Config/Definition/Builder/NodeBuilder.php]:
```suggestion
    protected (NodeDefinition&ParentNodeDefinitionInterface)|null $parent = null;
```

… in order to match the return type of `end()`. We could also think about adjusting the parameter type of `setParent()` accordingly.

File: src/Symfony/Component/Config/Definition/Builder/NodeDefinition.php
Comment (derrabus) [inline src/Symfony/Component/Config/Definition/Builder/NodeDefinition.php:38]: The return type of `end()` is wider. My gut feeling is that they should match.

**Comment by nicolas-grekas** [comment]
PR is ready for review. Needs #51121 to be :green_apple:

File: src/Symfony/Bundle/FrameworkBundle/Routing/Router.php
Comment (derrabus) [reply src/Symfony/Bundle/FrameworkBundle/Routing/Router.php]: Okay, that's what #51121 is for. Nevermind.

File: src/Symfony/Component/Config/Definition/Builder/NodeDefinition.php
Comment (nicolas-grekas) [reply src/Symfony/Component/Config/Definition/Builder/NodeDefinition.php:38]: All those classes in the return type of `end()` implement `NodeParentInterface` so we could reduce this to only it. I suppose the list is here to hint the IDE for autocompletion. Maybe we should keep only `NodeParentInterface` as native type and use `@return` for the rest?

The other strange thing is about the need to add `NodeInterface` to make this code work:
https://github.com/symfony/symfony/blob/1935af612953715faec8b69f2be70b20c7427bc5/src/Symfony/Component/Config/Definition/Builder/ArrayNodeDefinition.php#L352-L361

`end()` can't return an `ArrayNode` because of its return type. Either the return type is wrong, or what?
Up to investigate and fix this on a lower branch?

Comment (derrabus) [reply src/Symfony/Component/Config/Definition/Builder/NodeDefinition.php:38]: > All those classes in the return type of `end()` implement `NodeParentInterface` so we could reduce this to only it. I suppose the list is here to hint the IDE for autocompletion. Maybe we should keep only `NodeParentInterface` as native type and use `@return` for the rest?

Would work for me to unblock the PR.

> Up to investigate and fix this on a lower branch?

Maybe, but not right away. 🙈

Comment (nicolas-grekas) [reply src/Symfony/Component/Config/Definition/Builder/NodeDefinition.php:38]: > use @return for the rest

PR updated

**Review by derrabus — APPROVED** [review]

File: src/Symfony/Component/Config/Definition/Builder/NodeDefinition.php
Comment (stof) [reply src/Symfony/Component/Config/Definition/Builder/NodeDefinition.php:38]: The return type was not wrong. ArrayNode is not a NodeDefinition.

**Pull request merged**
